packages/protocol/contracts/layer1/based/TaikoInbox.sol

```diff
@@ -0,0 +1,805 @@
+            //  `keccak256(abi.encode("TAIKO_DIFFICULTY", block.number))`
+            info_ = BatchInfo({
+                txsHash: bytes32(0), // to be initialised later
+                //
+                // Data to build L2 blocks
+                blocks: params.blocks,
+                blobHashes: new bytes32[](0), // to be initialised later
+                extraData: bytes32(uint256(config.baseFeeConfig.sharingPctg)),
+                coinbase: params.coinbase,
+                proposedIn: uint64(block.number),
+                blobByteOffset: params.blobParams.byteOffset,
+                blobByteSize: params.blobParams.byteSize,
+                gasLimit: config.blockMaxGasLimit,
+                lastBlockId: 0, // to be initialised later
+                lastBlockTimestamp: lastBlockTimestamp,
+                //
+                // Data for the L2 anchor transaction, shared by all blocks in the batch
+                anchorBlockId: anchorBlockId,
+                anchorBlockHash: blockhash(anchorBlockId),
+                anchorInput: params.anchorInput,
+                baseFeeConfig: config.baseFeeConfig,
+                signalSlots: params.signalSlots
+            });
+
+            require(info_.anchorBlockHash != 0, ZeroAnchorBlockHash());
+
+            info_.lastBlockId = stats2.numBatches == config.forkHeights.pacaya
+                ? stats2.numBatches + uint64(params.blocks.length) - 1
+                : lastBatch.lastBlockId + uint64(params.blocks.length);
+
+            (info_.txsHash, info_.blobHashes) =
+                _calculateTxsHash(keccak256(_txList), params.blobParams);
+
+            meta_ = BatchMetadata({
+                infoHash: keccak256(abi.encode(info_)),
+                proposer: params.proposer,
+                batchId: stats2.numBatches,
+                proposedAt: uint64(block.timestamp)
+            });
+
+            Batch storage batch = state.batches[stats2.numBatches % config.batchRingBufferSize];
+
+            // SSTORE #1
+            batch.metaHash = keccak256(abi.encode(meta_));
+
+            // SSTORE #2 {{
+            batch.batchId = stats2.numBatches;
+            batch.lastBlockTimestamp = lastBlockTimestamp;
+            batch.anchorBlockId = anchorBlockId;
+            batch.nextTransitionId = 1;
+            batch.verifiedTransitionId = 0;
+            batch.reserved4 = 0;
+            // SSTORE }}
+
+            uint96 livenessBond =
+                config.livenessBondBase + config.livenessBondPerBlock * uint96(params.blocks.length);
+            _debitBond(params.proposer, livenessBond);
+
+            // SSTORE #3 {{
+            batch.lastBlockId = info_.lastBlockId;
+            batch.livenessBond = livenessBond;
+            batch._reserved3 = 0;
+            // SSTORE }}
+
+            stats2.numBatches += 1;
+            stats2.lastProposedIn = uint56(block.number);
+
+            emit BatchProposed(info_, meta_, _txList);
```

Comment:
Don't need the `_txList`, because we can [unpack](https://github.com/taikoxyz/taiko-mono/blob/966c5fc07d0a3de5ed8e719f9f188f53443a7fad/packages/taiko-client/driver/txlist_fetcher/calldata.go#L69) it from the params.